[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback

sinannasir opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41840 [NCCL] [For Test] In DDP's reducer merge `work` and `future_work`
* #42400 DDP communication hook: skip dividing grads by world_size if hook registered.
* **#42335 [NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback**

**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

sinannasir committed:
[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback

**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596). See the following approach:

**New design:**
1. As we create a `WorkNCCL` object inside `ProcessGroupNCCL::collective`, we also create a `FutureNCCL` (sub class of `Future`) associated with `WorkNCCL` and store it with the object. This future has a reference to the original work object.
2. The Future is marked as completed when its created (to allow for async execution of callbacks).
3. fut.wait() simply synchronizes the streams (synchronizeStreams() in Sinan's diff) and returns the result. This preserves the async execution model for CUDA.
4. When we add a callback to this future (.then()), we synchronizeStreams() and invoke the callback inline.

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596). See the following approach:

**New design:**
1. As we create a `WorkNCCL` object inside `ProcessGroupNCCL::collective`, we also create a `FutureNCCL` (sub class of `Future`) associated with `WorkNCCL` and store it with the object. This future has a reference to the original work object.
2. The Future is marked as completed when its created (to allow for async execution of callbacks).
3. fut.wait() simply synchronizes the streams (synchronizeStreams() in Sinan's diff) and returns the result. This preserves the async execution model for CUDA.
4. When we add a callback to this future (.then()), we synchronizeStreams() and invoke the callback inline.

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -125,9 +132,62 @@ class ProcessGroupNCCL : public ProcessGroup {
+    // entire operation to complete before returning.
+    void wait() override {
+      work_->wait();
+      markCompleted(at::IValue(*outputs_));
```
The issue here is that calling `completed()` on this Future will return True despite the work on the GPU not finishing. I think we should do the following:

1. Don't set `completed_ = true`.
2. Override `markCompleted()` and ensure it throws an Exception mentioning that the operation is not supported.
3. Override `value` and `constValue` to return `outputs_`.
4. Override `hasValue` to also return true.
5. Override `completed()` to return `work_->isCompleted()`.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
+            ``get_future`` retrieves a future associated with the completion of
+            ``c10d.ProcessGroup.work``. As an example, a future object can be set
```
nit: As an example, a future object can be **retrieved** by ...

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -125,9 +132,62 @@ class ProcessGroupNCCL : public ProcessGroup {
 
+  // FutureNCCL is a subclass of c10d ivalue's Future. FutureNCCL has a
+  // reference to WorkNCCL and NCCL collective's outputs. The goal is to
+  // use this class in getFuture API of WorkNCCL. Once NCCL kernel is done
```
> Once NCCL kernel is done executing, this future will have NCCL collective's outputs as its value

This isn't true? The future has a value even before the kernel is done executing?

Add a comment here mentioning this Future is mostly a wrapper to synchronize streams appropriately and it mostly enabled the async programming model of CUDA while trying to adhere to the Future interface.

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -125,9 +132,81 @@ class ProcessGroupNCCL : public ProcessGroup {
+
+    // Returns NCCL collective's outputs. This should only be used after NCCL
+    // kernel is done executing.
+    at::IValue value() override {
+      return outputs_;
+    }
+
+    const at::IValue& constValue() override {
+      return outputs_;
+    }
```
We probably need to call `work_->wait()` here as well to ensure we synchronize the streams correctly.

sinannasir replied:
After this change, I guess we don't need `work_->wait()` in `addCallback()` anymore:
https://github.com/pytorch/pytorch/blob/7352d01a980a8d724c061f2bfc0bc1fa4b164ac5/torch/lib/c10d/ProcessGroupNCCL.hpp#L187-L191

mrshenli replied:
I guess @pritamdamania87 is referring to `value()` and `constValue()`?

sinannasir replied:
Right, but if we call `work_->wait()` inside `value`, there is no need to call it before invoking the callback inline inside `addCallback()`.

mrshenli replied:
is that some newer version that hasn't been published yet? Looks like `work_->wait()` is not in `value()` in the current version?

sinannasir replied:
I'll update it soon.

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -3048,32 +3052,86 @@ def fut_then(fut):
+                # Multiply result by 10.
+                return [10 * t for t in fut.wait()]
+
+            return fut.then(fut_then)
```
Could we add another `.then` here to add more coverage? Ex: `return fut.then(fut_then).then(fooo)`

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -654,6 +657,30 @@ def _register_comm_hook(self, state: object, hook: callable):
+
+            >>> def encode_and_decode(state: object, bucket: dist.GradBucket): -> torch.futures.Future
+            >>>     encoded_tensors = encode(bucket.get_tensors()) # encode gradients
+            >>>     fut = process_group.allreduce(encoded_tensors).get_future()
```
I guess `encoded_tensors` here is not appropriate since `allreduce` takes only one tensor? We should probably do an allreduce for each tensor separately? Same for the allreduce example hook above.

sinannasir replied:
I think `process_group.allreduce` is fine because it can get a list of Tensors. The problem is with `dist.all_reduce` which just gets Tensors and not list of Tensors. 

@mrshenli also mentioned renaming this function as `get_replicas` for better engineering. Maybe we have both, and `get_tensors` returns Tensors but raises exception if there are multiple replicas? 
My plan is to address the minor API changes (like this example and skipping division by worldsize) in a stacked diff.

sinannasir replied:
Please see the new `get_tensor` method in GradBucket. I think it makes sense:
https://github.com/pytorch/pytorch/blob/f421853b0b950c2b0bae0100558f39bc3082d8c0/torch/csrc/distributed/c10d/comm.cpp#L90-L96

pritamdamania87 replied:
I think we should have only one. I'm fine with calling it `get_replicas`, but the docs should clearly mention that each tensor in the list refers to the replica on each device. Also, mention that we will have multiple replicas only in the case of single process multiple device mode. In the single process single device mode, this list would consist of only a single tensor.

sinannasir replied:
I'll follow the recommended change here, but I didn't get how to `allreduce` each tensor separately. How can I combine those works into one Future?

sinannasir replied:
Reverted `getTensor`

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -374,11 +374,11 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   }
 
-  bool hasValue() const {
+  virtual bool hasValue() const {
```
cc @wanchaol, is it OK to make these functions virtual?

mrshenli commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -113,14 +113,18 @@ c10::intrusive_ptr<torch::jit::Future> PythonCommHook::runHook(
+  // Since we have a Python hook, future_value can be a PyObject.
+  if (future_value.isPyObject()) {
+    // We first convert it to an IValue that contains a TensorVector.
+    py::object obj =
```
do we need GIL to protect these or is this protected on all callsites?

sinannasir replied:
I think it should be fine without GIL. @pritamdamania87 Could you comment on this?

mrshenli replied:
Is the GIL held by the callsite? If no, we will need a GIL here.

sinannasir replied:
This is just called by `reducer::finalize_backward` so i believe GIL wasn't held anywhere before the call. I included GIL for correctness.

mrshenli commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -113,14 +113,18 @@ c10::intrusive_ptr<torch::jit::Future> PythonCommHook::runHook(
+  if (future_value.isPyObject()) {
+    // We first convert it to an IValue that contains a TensorVector.
+    py::object obj =
+        py::reinterpret_borrow<py::object>(future_value.toPyObject());
```
Any reason for not directly using `torch::jit::toPyObject(future_value)`?

mrshenli commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -21,6 +21,7 @@ void broadcast_coalesced(
 class GradBucket {
  public:
   explicit GradBucket(std::vector<at::Tensor> tensors);
+
```
unintentional new line?

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+              -> std::shared_ptr<jit::PythonFutureWrapper> {
+            return std::make_shared<jit::PythonFutureWrapper>(work.getFuture());
+          },
+          py::call_guard<py::gil_scoped_release>(),
```
`getFuture` seems to be very light-weight, if so, we don't need to release the GIL so that we can avoid additional context switch.

sinannasir replied:
makes sense.

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+          R"(
+            ``get_future`` retrieves a future associated with the completion of
+            ``c10d.ProcessGroup.work``. As an example, a future object can be
+            retrieved by `fut = dist.allreduce(tensors).get_future()`.
```
double quote

mrshenli replied:
I believe you mean `dist.all_reduce(tensor).get_future()`

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            ``c10d.ProcessGroup.work``. As an example, a future object can be
+            retrieved by `fut = dist.allreduce(tensors).get_future()`.
+
+            In the example above, if `dist.allreduce` work was done on GPU,
```
double quote all code in the doc.

mrshenli replied:
allreduce -> all_reduce

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+
+            In the example above, if `dist.allreduce` work was done on GPU,
+            `fut.wait()` would return after synchronizing the correct GPU streams
+            to ensure we can have async CUDA execution and it does not wait for
```
async -> asynchronous

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            In the example above, if `dist.allreduce` work was done on GPU,
+            `fut.wait()` would return after synchronizing the correct GPU streams
+            to ensure we can have async CUDA execution and it does not wait for
+            the entire operation to complete on GPU. If NCCL_BLOCKING_WAIT is enabled,
```
> `fut.wait()` would return after synchronizing the correct GPU streams to ensure we can have async CUDA execution and it does not wait for the entire operation to complete on GPU.

what does it referring to here? And what does `synchronizing` mean here? I am assuming `synchronizing` does not mean blocking here? If so, please expand this to make that clear. Only saying `synchronizing` is a little confusing, as people might think it's similar to `torch.cuda.synchronize()`, which is blocking

sinannasir replied:
I think we wanna give a sense of what we did in NCCL here. My understanding in NCCL `synchronize` just refers to `cudaEvents_[i].block(currentStream);` cc @pritamdamania87 any thoughts?

mrshenli replied:
Yep, I actually understand what you mean here. Trying to review from a user's perspective, and assuming the user is not familiar `ProcessGroupNCCL` implementation.

pritamdamania87 replied:
For users familiar with synchronization between streams on CUDA, shouldn't this be clear that we're synchronizing the correct streams here? We could expand this doc to mention that NCCL uses its own pool of streams and pytorch has its own default streams per device. Synchronizing here means that we're synchronizing the appropriate NCCL streams with the device streams such that subsequent operations on the return value are processed correctly by the default device stream.

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+
+            Note that `fut.done()` actually returns if the work was completed on the GPU.
+
+            .. warning ::
```
Add one more warning to highlight how this is different from `torch.futures.Future`.

sinannasir replied:
I updated the description. I'm not sure about the differences between `toch_C.Future` and `torch.futures.Future`. I asked that question in a separate thread.

mrshenli replied:
This is actually not about the difference between `torch.futures.Future` and `torch._C.Future`, instead, we want to distinguish `torch.futures.Future` from the Future type returned by `get_future`.  We are just using the `torch._C.Future` type to infer that "this is an internal type, please read the limitations and use it with caution".

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            `fut.wait()` would return after synchronizing the correct GPU streams
+            to ensure we can have async CUDA execution and it does not wait for
+            the entire operation to complete on GPU. If NCCL_BLOCKING_WAIT is enabled,
+            in that case, it would wait for the entire operation to complete before
```
Let's add an example below and use code + comments to explain how `NCCL_BLOCKING_WAIT` changes the behavior.

sinannasir replied:
Based on the internal discussion, I think it is okay to leave this as is. just in case cc'ing @pritamdamania87

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            the entire operation to complete on GPU. If NCCL_BLOCKING_WAIT is enabled,
+            in that case, it would wait for the entire operation to complete before
+            returning. In addition, if a callback function was added by `fut.then()`,
+            it will synchronize the appropriate streams and invoke the callback inline.
```
hmm, is this second "synchronize" means blocking here?

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            returning. In addition, if a callback function was added by `fut.then()`,
+            it will synchronize the appropriate streams and invoke the callback inline.
+
+            Note that `fut.done()` actually returns if the work was completed on the GPU.
```
This is as expected, right?

sinannasir replied:
I think so, this is what @rohan-varma mentioned in the group chat.

mrshenli replied:
I see. Then let's remove "actually".

mrshenli commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -55,8 +55,11 @@ bool ProcessGroup::Work::wait(std::chrono::milliseconds timeout) {
-void ProcessGroup::Work::abort() {
-  TORCH_CHECK(false, "ProcessGroup::Work::abort not implemented.")
+void ProcessGroup::Work::abort(){
+    TORCH_CHECK(false, "ProcessGroup::Work::abort not implemented.")}
```
wrong indent and spacing, please use clang-format to format code.

sinannasir replied:
That was done by VS code. Actually, when I do
```
void ProcessGroup::Work::abort(){
  TORCH_CHECK(false, "ProcessGroup::Work::abort not implemented.")}
```
VS Code gives `clang-format` warning. Maybe it is a bug with VS code's language service.

mrshenli replied:
what about the following?

```c++
void ProcessGroup::Work::abort() {
  TORCH_CHECK(false, "ProcessGroup::Work::abort not implemented.");
}
```

sinannasir replied:
That worked, thanks!

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -125,9 +132,81 @@ class ProcessGroupNCCL : public ProcessGroup {
     friend class ProcessGroupNCCL;
   };
 
+  // FutureNCCL is a subclass of c10d ivalue's Future. The goal is to
```
remove "c10d", ivalue is in JIT

mrshenli commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -89,6 +89,10 @@ class ProcessGroup {
 
     virtual void abort();
 
+    // Returns a Future object that will be marked as completed once work is
```
Isn't the future marked as completed on construction? I understand that the `completed()` API will return work's completed status. If this is correct let's make it explicit in the comment.

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -125,9 +132,81 @@ class ProcessGroupNCCL : public ProcessGroup {
+    // FutureNCCL has a value that was set in its constructor as NCCL
+    // collective's outputs.
+    bool hasValue() const override {
+      return true;
```
do we need to wait in here as well?

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -633,6 +633,9 @@ def _register_comm_hook(self, state: object, hook: callable):
                              c10d reducer would call this hook and use the tensors returned
                              by the Future and copy grads to individual parameters.
 
+                             We also provide an API called "get_future" to retrieve a future
```
Add a warning to highlight that this future is different from `torch.futures.Future` and explain the differences.

mrshenli replied:
use `` instead of " to wrap code.

sinannasir replied:
@mrshenli Could you elaborate more on the differences between `torch.futures.Future` and `torch._C.Future`? The former is used in jit and latter is just the `at::ivalue`'s Future?

mrshenli replied:
`torch.futures.Future` and `torch._C.Future` is roughly the same, where the former is sheer wrap of the latter. The concern is actually that the return future type of `get_future` does not behave the same way as `torch.futures.Future`, so that we need to highlight that.

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -654,6 +657,30 @@ def _register_comm_hook(self, state: object, hook: callable):
+        Example::
+            Below is an example of a simple allreduce hook.
+
+            >>> ddp._register_comm_hook(state = None, hook = allreduce)
```
don't you need to flip the order of this the allreduce below, otherwise this would throw "NameError: name 'allreduce' is not defined"

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -654,6 +657,30 @@ def _register_comm_hook(self, state: object, hook: callable):
+
+            >>> ddp._register_comm_hook(state = None, hook = allreduce)
+
+            >>> def allreduce(state: object, bucket: dist.GradBucket): -> torch.futures.Future
```
hmm, the return type is not `torch.futures.Future`, as they are different, right?

Actually what does `type(work.get_future())` return?

sinannasir replied:
It returns `torch._C.Future`

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -654,6 +657,30 @@ def _register_comm_hook(self, state: object, hook: callable):
+
+            >>> ddp._register_comm_hook(state = None, hook = encode_and_decode)
+
+            >>> def encode_and_decode(state: object, bucket: dist.GradBucket): -> torch.futures.Future
```
`GradBucket` is not in master do, and it's also not available at `c10d/init.cpp`. Could you please add docs for that? And since `_register_comm_hook` is a private API, I guess we need to make `GradBucket` and `get_tensors` private as well?

sinannasir replied:
@pritamdamania87 I believe I asked a similar question before. Could you comment on this?

pritamdamania87 replied:
GradBucket is actually in c10d/init.cpp https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/init.cpp#L139, we can probably just add some docs there.

Agree on making `GradBucket` private, I guess `get_tensors` would be private since the class `GradBucket` is private as well.

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -3048,32 +3052,86 @@ def fut_then(fut):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        def allreduce_hook(state: object, bucket: dist.GradBucket) -> torch.futures.Future:
```
I see, maybe both `torch.futures.Future` and `.get_future()` will return `torch.jit.Future`. Can we modify the return type to `torch._C.Future` for now, and we can change it back to `torch.futures.Future` when it really behaves the same way.

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -168,25 +167,23 @@ class ProcessGroupNCCL : public ProcessGroup {
     void addCallback(std::function<void(void)> callback) override {
-      work_->wait();
-
       // Run callback inline after synchronizeStreams.
```
nit: This comment is not accurate anymore.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -142,6 +142,10 @@ PyObject* c10d_init(PyObject* _unused) {
       .def(
           "get_tensors",
           &::c10d::GradBucket::getTensors,
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "get_tensor",
```
I think we should have just one API that returns the entire list of tensors as I mentioned in a separate comment.

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -125,9 +132,78 @@ class ProcessGroupNCCL : public ProcessGroup {
+    // Callbacks return a Future (not FutureNCCL).
+    void addCallback(std::function<void(void)> callback) override {
+      // Run callback inline after synchronizeStreams.
+      callback();
```
I might have missed some context but why don't we call `wait()` here (or ensure wait() has been called before this)? If the callback is invoked inline and we have not synchronized streams, then as I understand it the callback could do some work on the default stream using the result of the future, but the result won't actually have the value since the async op on the nccl stream did not complete?

sinannasir replied:
My understanding is `callback` will be blocked by `fut.value` when it tries to get the value.

mrshenli replied:
> My understanding is that callback has to call fut.value to get the value and it should synchronize the streams.

Is this always true? what if a user adds a callback for profiling reason which does not access the value? Or does our code access the value somewhere?

sinannasir replied:
Yeah, if user doesn't call value or wait  inside callback. It won't work as expected. :) 
Should we add a warning about that? Or should revert back to the version where we wait before calling callback() inline, and do `assert(work->isCompleted())` instead of `work->wait()` inside `value` and `constvalue`?
cc'ed @pritamdamania87

pritamdamania87 replied:
Let's just add a `wait()` here since in the case of profiling or something this wouldn't work as expected.

sinannasir replied:
Makes sense. Then, for efficiency I'm removing `wait` inside `value` and just `assert(work->isCompleted()) before returning value.

sinannasir replied:
Makes sense. Then, for efficiency I'm removing `wait` inside `value`. It doesn't make sense to call wait multiple times for the then callback case.

pritamdamania87 replied:
Let's keep the wait inside value as well :) Its unlikely, but it is possible that the user calls wait() on one thread and value() on another thread. As a result, it's not guaranteed that the streams have been synchronized.

sinannasir replied:
Sure,

rohan-varma commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -87,6 +87,14 @@ const std::vector<at::Tensor>& GradBucket::getTensors() {
+  TORCH_CHECK(
+      tensors_.size() == 1,
+      "GradBucket::getTensor can only be used when tensors "
+      "contain bucket contents tensor for just one replica.")
```
Does this mean that the comm. hook does not work when modules are replicated across GPU in a single process (single-process, multi-device)? This is fine as we're retiring support for that, but would be good to confirm and maybe clarify in docs (cc @mrshenli )

sinannasir replied:
I removed `getTensor` after the thread in https://github.com/pytorch/pytorch/pull/42335#discussion_r463769165  
I think `process_group.allreduce()` gets list of Tensors and it should support the multiple replica case. 

Looks like we need to `allreduce` separately for each replica in the case of `dist.all_reduce()`. Since we need to combine all these `allreduce` into one Future, I'm not really sure how the hook will look like for that case though.

mrshenli replied:
> I'm not really sure how the hook will look like for that case though.

in that case, we can either:

1. add a warning to state that the hook does not work with SPMD mode
2. in `_register_comm_hook` detect if it is SPMD, if yes, error out.

sinannasir replied:
@mrshenli but `process_group.allreduce()` gets list of Tensors, so it should work with SPMD mode, right?

mrshenli replied:
@sinannasir yep, I agree. But as a framework, we don't promise on features without test coverage. If you plan to imply it would work for both SPMD DDP and SPSD DDP, we will need to add tests for both modes.

pritamdamania87 replied:
@sinannasir To keep this PR simple, let's just error out in the SPMD mode and create a gh issue to address this later.

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -622,7 +622,9 @@ def _register_comm_hook(self, state: object, hook: callable):
                             next in GossipGrad etc.
             hook (callable): is defined as:
-                             hook(state: object, bucket: dist.GradBucket) -> torch.futures.Future:
+                             hook(state: object, bucket: dist._GradBucket) -> torch.futures.Future:
+                             or
+                             hook(state: object, bucket: dist._GradBucket) -> torch._C.Future:
```
My feeling is that we should probably have only one of these, although I'll let @mrshenli confirm this.

mrshenli replied:
Agree, we should can just say it returns `torch.futures.Future` here, as this is a requirement for user code. 

Then in `get_future` API doc, we highlight that it does not return `torch.future.Future` and describe all the differences there. And also mention if the application would like to return `get_future()`'s future from `hook(...)`, what they need to be aware.

pritamdamania87 reviewed (commented):
Looks good to me overall, will let @mrshenli stamp this PR once his comments are addressed.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -136,7 +137,7 @@ PyObject* c10d_init(PyObject* _unused) {
+  shared_ptr_class_<::c10d::GradBucket>(module, "_GradBucket")
       .def(py::init<std::vector<Tensor>&>(), py::arg("tensors"))
       .def(
           "get_tensors",
```
Can we add documentation to this method (similar to the documentation in the C++ getTensors method) explaining the replicas?

mrshenli commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -113,14 +113,17 @@ c10::intrusive_ptr<torch::jit::Future> PythonCommHook::runHook(
+  // Since we have a Python hook, future_value can be a PyObject.
+  if (future_value.isPyObject()) {
+    // We first convert it to an IValue that contains a TensorVector.
+    py::object obj = torch::jit::toPyObject(future_value);
```
This is a correctness comment: if the call site is not always holding GIL, we will need GIL here.

pritamdamania87 replied:
Yes, I think we need GIL here since the callsite is c10d/reducer.cpp which doesn't hold GIL.

mrshenli commented:
Is this test error introduced by this PR or is it already on master?

```
Aug 01 05:50:08 [W CudaIPCTypes.cpp:22] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
Aug 01 05:50:08 Traceback (most recent call last):
Aug 01 05:50:08   File "test/run_test.py", line 752, in <module>
Aug 01 05:50:08     main()
Aug 01 05:50:08   File "test/run_test.py", line 741, in main
Aug 01 05:50:08     raise RuntimeError(err)
Aug 01 05:50:08 RuntimeError: distributed/test_c10d_spawn failed! Received signal: SIGSEGV
```

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

sinannasir commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -125,9 +132,77 @@ class ProcessGroupNCCL : public ProcessGroup {
+
+    // FutureNCCL has a value that was set in its constructor as NCCL
+    // collective's outputs.
+    bool hasValue() const override {
```
@pritamdamania87 @mrshenli I followed @wanchaol's comment and return `completed` inside `hasValue`. The only problem is outputs could be ready and accessed by reducer or callback before `completed` is true since it checks for `work->isCompleted()`. I think that's fine though.

mrshenli replied:
> The only problem is outputs could be ready and accessed by reducer or callback before completed is true since it checks for work->isCompleted(). 

Could you please elaborate? Why the outputs can be ready before the NCCL work is completed?

pritamdamania87 replied:
@wanchaol @mrshenli FutureNCCL is a wrapper around c10d work to ensure we can adhere to the Future's API and at the same time support the async model that CUDA provides. In the current version, you can call `value()` before the work has actually completed on the GPU and use that value in another operation. This is possible due to the async nature of CUDA and the fact that we synchronize the streams correctly. Since, you can always call `value()` irrespective of whether or not the work has completed on the GPU, I think for consistency `hasValue()` should always return True.

sinannasir replied:
> > The only problem is outputs could be ready and accessed by reducer or callback before completed is true since it checks for work->isCompleted().
> 
> Could you please elaborate? Why the outputs can be ready before the NCCL work is completed?

@mrshenli By default `NCCL_BLOCKING_WAIT=0`, so `work->wait()` will just return after synchronizing streams, whereas `work->isCompleted()` still checks for cuda success.

sinannasir replied:
@mrshenli @wanchaol reverted the change on `hasValue` after @pritamdamania87's comment. I think it makes sense to simply return true, since we set the value right at the constructor and always wait inside `FutureNCCL::value`.

sinannasir commented:
> Is this test error introduced by this PR or is it already on master?
> 
> ```
> Aug 01 05:50:08 [W CudaIPCTypes.cpp:22] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
> Aug 01 05:50:08 Traceback (most recent call last):
> Aug 01 05:50:08   File "test/run_test.py", line 752, in <module>
> Aug 01 05:50:08     main()
> Aug 01 05:50:08   File "test/run_test.py", line 741, in main
> Aug 01 05:50:08     raise RuntimeError(err)
> Aug 01 05:50:08 RuntimeError: distributed/test_c10d_spawn failed! Received signal: SIGSEGV
> ```

Just rebased and will see whether it goes away. If not, will have a deeper look.

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -633,27 +633,60 @@ def _register_comm_hook(self, state: object, hook: callable):
+            Below is an example of a Parallel SGD algorithm where gradients are encoded before
+            allreduce, and then decoded after allreduce.
+
+            >>> def encode_and_decode(state: object, bucket: dist._GradBucket): -> torch._C.Future
```
Does `.then()` return `FutureNCCL` or `torch.futures.Future`? I assume it's the latter? In this case, we the return type should be `torch.futures.Future`? I guess we only need to when `torch._C.Future` as the return type when the hook function directly returns `FutureNCCL`.

sinannasir replied:
I tried the following:
```c++
        def allreduce_then_hook(state: object, bucket: dist._GradBucket) -> torch._C.Future:
            tensors = bucket.get_tensors()#[t / self.world_size for t in bucket.get_tensors()]
            fut = process_group.allreduce(tensors).get_future()

            def mult(fut):
                # Multiply result by 10.
                return [10 * t for t in fut.wait()]

            def div(fut):
                print(type(fut))
                # Divide result by 2.
                return [0.5 * t for t in fut.wait()]

            return fut.then(mult).then(div)
```
`print(type(fut))` prints `torch._C.Future`. I think any future that has something to do with C++ no matter its either `at::ivalue::Future ` or `FutureNCCL` will be this type in Python. cc'ed @pritamdamania87

mrshenli replied:
It is true that all these future types are some wrapper/derived type of `torch._C.Future`, but we were only intended to use that to refer to `FutureNCCL`. The original thought is that we have the following mapping:

`torch._C.Future` -> `FutureNCCL`
`torch.futures.Future` -> all other futures

Since I cannot explain this clearly between us during the past few rounds of discussions, I guess this might be even harder to convey accurately to users. Does the following make more sense to you?

1. Create a new pybind type that exposes `FutureNCCL` as `torch.distributed.FutureNCCL` to Python.
2. Explicitly mark the return type of `get_future()` as `torch.distributed.FutureNCCL`.
3. Use `torch.futures.Future` to refer to all other future types.
4. In the API doc of `FutureNCCL`, explain the difference between `torch.distributed.FutureNCCL` and `torch.futures.Future`

sinannasir replied:
@mrshenli Please see the latest version. I hope I covered everything based on our discussion offline.

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -633,27 +633,60 @@ def _register_comm_hook(self, state: object, hook: callable):
 
+                             We also provide an API called ``get_future`` to retrieve a
+                             Future associated with the completion of c10d.ProcessGroupNCCL.work.
+                             Note that this API will return a toch._C.Future which is an
```
I might miss-communicated in my previous comments. This `hook()` should return `torch.futures.Future`, as this is a requirement for users. The `get_future()` API returns `torch._C.Future`, as we should clarify how it is different from `torch.futures.Future` in the API doc of `get_future()`.

Think from a user's perspective. This API doc should tell a user how to write a hook. If we say the hook should return a `torch._C.Future`, then the user might be confused about what is `torch._C.Future`? how can I construct `torch._C.Future`? Why this should return `torch._C.Future`?

The reason we let `get_future()` return a `torch._C.Future` is not because `hook()` should return `torch._C.Future`. It is only a compromise we made for NCCL operations, and the `hook()` API is serving a broader range of applications than NCCL ops. So the limitation and special-casing should be on `get_future()` instead of on `hook()`.

sinannasir replied:
Referring to the thread in https://github.com/pytorch/pytorch/pull/42335#discussion_r465207732 

By saying "this API will return a toch._C.Future", I was meaning "get_future" API. Revised the explanation. Hope looks better now.

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +708,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            asynchronous CUDA execution and it does not wait for the entire operation
+            to complete on GPU. If NCCL_BLOCKING_WAIT is enabled, in that case, it would
+            wait for the entire operation to complete before returning. In addition,
+            if a callback function was added by `fut.then()`, it will wait until workNCCL's
```
double quote

mrshenli replied:
you mean WorkNCCL?

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +708,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            if a callback function was added by `fut.then()`, it will wait until workNCCL's
+            wait returns and invoke the callback inline.
+
+            Note that `fut.done()` returns if the work was completed on the GPU.
```
double quote

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +708,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            backend, ``fut.wait()`` would return after synchronizing the appropriate
+            NCCL streams with PyTorch's default device streams to ensure we can have
+            asynchronous CUDA execution and it does not wait for the entire operation
+            to complete on GPU. If NCCL_BLOCKING_WAIT is enabled, in that case, it would
```
double quote `NCCL_BLOCKING_WAIT `

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +708,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+          },
+          R"(
+            ``get_future`` retrieves a future associated with the completion of
+            ``c10d.ProcessGroup.work``. As an example, a future object can be
```
`c10d.ProcessGroup.work` is not visible from Python. You mean `ProcessGroup::Work`?

mrshenli replied:
Maybe more specifically, it should be `ProcessGroupNCCL::WorkNCCL`

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +708,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            return std::make_shared<jit::PythonFutureWrapper>(work.getFuture());
+          },
+          R"(
+            ``get_future`` retrieves a future associated with the completion of
```
Replace this paragraph with:

```
Returns:
    A ``torch._C.Future`` object which is associated with the completion of the 
    ``ProcessGroupNCCL::WorkNCCL``. As an example, a future object can be
    retrieved by ``fut = process_group.allreduce(tensors).get_future()``.
```

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +708,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            ``c10d.ProcessGroup.work``. As an example, a future object can be
+            retrieved by ``fut = process_group.allreduce(tensors).get_future()``.
+
+            In the example above, if ``allreduce`` work was done on GPU using NCCL
```
Add a sentence to the front of this paragraph:

```
The ``torch._C.Future`` object returned by this API can be used in 
``DistributedDataParallel._register_comm_hook``, but it is subject to
some subtle differences compared to ``torch.futures.Future`` due
to compromises made for performance reasons. ...
```

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,7 +708,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
+            retrieved by ``fut = process_group.allreduce(tensors).get_future()``.
+
+            In the example above, if ``allreduce`` work was done on GPU using NCCL
+            backend, ``fut.wait()`` would return after synchronizing the appropriate
```
> if ``allreduce`` work was done on GPU using NCCL

Does other backend support this API? If not, let's be explicit.

sinannasir replied:
It currently only supports NCCL. I revised it and mention API's warning inside this paragraph as well.

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -136,12 +137,19 @@ PyObject* c10d_init(PyObject* _unused) {
-          py::call_guard<py::gil_scoped_release>());
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
+            ``get_tensors`` returns a list of torch.Tensor. Each tensor in the list
```
Use doc format:

```
Returns:
    A list of ``torch.Tensor``....
```

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -3008,37 +3011,42 @@ def test_ddp_comm_hook_future_passing_cpu(self):
+
+    def _simple_hook(
+        self, state: object, bucket: dist._GradBucket
+    ) -> torch.futures.Future:
```
Do we also have tests to cover the case where the hook directly returns the `FutureNCCL` acquired from `.get_future()`? Instead of chaining another `.then()`?

sinannasir replied:
Yes we do. Please see `test_ddp_comm_hook_allreduce_hook_nccl`:
https://github.com/pytorch/pytorch/blob/b49167474549b61f5fb554bfac6ef681e9b41654/test/distributed/test_c10d.py#L3105-L3106

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Let's land

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -756,7 +756,7 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
 c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupNCCL::WorkNCCL::
     getFuture() {
-  return futureWork_;
+  return c10::make_intrusive<FutureNCCL>(shared_from_this(), outputs_);
```
nit: Add a comment here about the circular reference and why we need to create a new object here.

mrshenli reviewed (approved):
Thanks for fixing the circular reference. LGTM! Please address @pritamdamania87's comments before landing. Thanks!

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture() without cudaStreamAddCallback"


**Main goal:** For DDP communication hook, provide an API called "get_future" to retrieve a future associated with the completion of c10d.ProcessGroupNCCL.work. Enable NCCL support for this API in this diff.

We add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.

We no longer consider a design that involves cudaStreamAddCallback which potentially was causing performance regression in [#41596](https://github.com/pytorch/pytorch/pull/41596).

Differential Revision: [D22833298](https://our.internmc.facebook.com/intern/diff/D22833298/)

[ghstack-poisoned]
